Fkm/form submission

Gabrimari opened the pull request:
Add endpoint for form submission to the database.

Gabrimari committed:
Store session identifier in payment instead of cookie.

Gabrimari committed:
Update frontend to use new session identifiers.

Gabrimari committed:
Update views such that email address is handled by reserve ticket instead.

Gabrimari committed:
Update docs with new endpoint information.

Gabrimari committed:
Merge branch 'fkm/main' into fkm/form-submission

Gabrimari committed:
Add structure for submit-form endpoint.

Gabrimari committed:
Add some tests for submit-form view.

Gabrimari committed:
Merge branch 'fkm/form-submission' of github.com:Fysiksektionen/bittan into fkm/form-submission

Gabrimari committed:
Add tests for submit-form endpoint.

Gabrimari committed:
Add submit form view.

Gabrimari committed:
Add first come first serve field to chapter event.

Gabrimari committed:
submit_form: Check if Payment in correct status

Gabrimari committed:
Add documentation for submit form

Gabrimari committed:
Merge branch 'fkm/main' into fkm/form-submission

Gabrimari committed:
Change tests to properly access an event with form.

Gabrimari committed:
Divide submit_form view into type hinted functions to improve readability.

Gabrimari committed:
Add tests and functionality to update form answers in submit_form view.

JonathanGodar commented on backend/bittan/bittan/views/submit_form.py:
```diff
@@ -0,0 +1,189 @@
+    "QuestionNotFound": status.HTTP_404_NOT_FOUND,
+    "QuestionOptionNotFound": status.HTTP_404_NOT_FOUND
+}
+error_helper = lambda msg: Response(msg, CODE_MAPPINGS[msg])
```
Skriv om som funktion?

Gabrimari replied:
Resolved in 1b0f010be5d277a05b7bdaba3ac399ae953b9420

JonathanGodar commented on backend/bittan/bittan/views/submit_form.py:
```diff
@@ -0,0 +1,189 @@
+    session_id = serializers.CharField(required=True)
+    form_data = serializers.ListField(child=QuestionSerializer(), required=True)
+
+def validate_payment(
```
Get payment?

Gabrimari replied:
Resolved in 1b0f010be5d277a05b7bdaba3ac399ae953b9420

JonathanGodar commented on backend/bittan/bittan/views/submit_form.py:
```diff
@@ -0,0 +1,189 @@
+	
+	return payment
+
+def validate_event(
```
Kanske tydligare namn?

Gabrimari replied:
Resolved in 1b0f010be5d277a05b7bdaba3ac399ae953b9420

JonathanGodar commented on backend/bittan/bittan/views/submit_form.py:
```diff
@@ -0,0 +1,189 @@
+		if not chapter_event.question_set.filter(pk=qs["question_id"]).exists():
+			return error_helper("QuestionNotFound")
+		
+		q_db = chapter_event.question_set.get(pk=qs["question_id"])
```
Kommer skicka en jäkla massa DB requests men spelar förmodligen ingen roll

Gabrimari replied:
Resolved in 18cb38d0bae699c166f4df123e6555ba8d1a7f7a

Now fetches all database data regarding the questions at the start of the function. Should reduce the number of queries to a total of 2. Drawback is that the question data is cached during the entire function. If changes are made to the questions in database mid function then some problems might arise. Since questions as a rule of thumb not should be changed when an event is live it should not provide major issues.

JonathanGodar commented on backend/bittan/bittan/views/submit_form.py:
```diff
@@ -0,0 +1,189 @@
+	if isinstance(payment, Response):
+		return payment
+    
+    # Only supports forms if there are one ticket (for now). 
```
Kracha om fler än en ticket?

Gabrimari replied:
Resolved in 1b0f010be5d277a05b7bdaba3ac399ae953b9420

JonathanGodar commented on backend/bittan/bittan/views/submit_form.py:
```diff
@@ -0,0 +1,189 @@
+		r["question_id"]: r for r in response_data["form_data"]
+	}
+
+	q_errors = validate_questions(form_data_map, chapter_event)	
```
Fett najs :)

JonathanGodar commented on backend/bittan/bittan/views/submit_form.py:
```diff
@@ -0,0 +1,189 @@
+	if q_errors:
+		return q_errors
+
+	# Update everhything at once (!?) if the session is valid. 
```
!? hmmm

Gabrimari replied:
Resolved in 1b0f010be5d277a05b7bdaba3ac399ae953b9420

Gabrimari committed:
Resolve reviews

Errorhandler as function
Rename validate_payment
Rename validate_event
Error on more thatn 1 ticket
Update comment !?

Gabrimari committed:
Reduce database requests in submit_form validate_questions

Gabrimari committed:
Remove stray print statement and fix typo

Gabrimari committed:
Only add answered quesitons

Now skips over any non-manadatory questions that has been left empty in the answer.

JonathanGodar commented on backend/bittan/bittan/views/submit_form.py:
```diff
@@ -116,29 +126,32 @@ def validate_questions(
 
 @api_view(["POST"])
 def submit_form(request: Request) -> Response:
+	print("Got to the view")
```
Ta bort debug print?
